### PR TITLE
Add user provisioning overlay methods

### PR DESCRIPTION
**Pull request by skyeturriff**
### Intent

OS companion to https://github.com/rstudio/rstudio-pro/pull/5057

Once both these PRs are approved, I will merge this one and sync to Pro, and then handle any conflicts before merging the Pro PR.

### Approach

The Pro and OS versions of `ServerAuthHandler.cpp` had diverged a bit; this PR attempts to align that file to only include shared code and move non-shared code into `ServerAuthHandlerOverlay.cpp`.

Adds the new `ServerAuthHandlerOverlay.cpp` and header files, with the open source implementation of the overlay methods updated/added for the Pro user provisioning feature. Many of these are just stubs, with the exception of `addUserToDatabase` and `getUsernameDbColumnName` which handle the difference in the username column.

### Automated Tests

If I can help with this, please advise!

### QA Notes

These changes touch code that handles:
- user login
- getting users, when listing users to share a project with, or via the `rstudio-server list-users` command
- adding users, when using the `rstudio-server add-user` command

### Documentation

n/a

### Checklist

- [ ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [ ] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [ ] This PR passes all local unit tests

### CODE REVIEW

**Review by zachhannum — APPROVED** [review]
Just a few comments, but otherwise this LGTM!

File: src/cpp/server/auth/ServerAuthHandlerOverlay.cpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+OverlayResult addUser(boost::asio::io_service& ioService,
+                      const std::string& username,
+                      bool isAdmin)
+{
+   return std::make_tuple(Success(), false);
+}
+
+OverlayResult addUserToDatabase(const boost::shared_ptr<IConnection>& connection,
```
[the remark this answers is not in the document]
Comment (skyeturriff) [reply src/cpp/server/auth/ServerAuthHandlerOverlay.cpp]: Good point! Fixed!

File: src/cpp/server/auth/ServerAuthHandlerOverlay.cpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+bool isUserProvisioningEnabled()
+{
+   return false;
+}
+
+std::string getUsernameDbColumnName()
+{
+   return "user_name";
+
```
[the remark this answers is not in the document]
Comment (skyeturriff) [reply src/cpp/server/auth/ServerAuthHandlerOverlay.cpp]: Thank you!

File: src/cpp/server/auth/ServerAuthHandler.cpp
[the remark this answers is not in the document]
Comment (skyeturriff) [reply src/cpp/server/auth/ServerAuthHandler.cpp:1]: Done!

**Review by MariaSemple — COMMENTED** [review]
This looks good, but shouldn't we just modify the OS database to include the same columns as the pro database? Even if most of them never get used in OS, I think it will be cleaner for long term maintenance

**Comment by zachhannum** [comment]
I agree updating the OS database would clean up some of the overlay functionality. Just to be clear, are you suggesting we update the OS `licensed_users` table to include *all* of the columns that Pro has, or just the ones that overlap in functionality in the auth handler code? i.e. `posix_name`, `display_name`, etc.

**Comment by MariaSemple** [comment]
All - so there's no difference between the OS database and the Pro database

**Comment by zachhannum** [comment]
> no difference between the OS database and the Pro database

So do we want to just completely align the OS and Pro schema across _all_ tables in the database?

There are also a number of triggers/trigger functions that were introduced that we'd want to make sure get added if we want OS and Pro databases to be compatible.

**Comment by kevinushey** [comment]
Checking in; is this ready for merge?

**Review by zachhannum — APPROVED** [review]
I think this is ready!

**Review by MariaSemple — APPROVED** [review]
LGTM!

File: src/cpp/server/auth/ServerAuthHandlerOverlay.cpp [hunk 3]
```diff
@@ -125,7 +125,8 @@ Error checkForUninitializedUsername(const boost::shared_ptr<IConnection>& connec
                                     const system::User& user,
                                     std::string* pUsername)
 {
-   // we don't expect the username to be uninitialized in open source
+   // we don't expect the username to be uninitialized in open source, so just set it and return
+   *pUsername = row.get<std::string>(0);
```
Comment (zachhannum) [inline src/cpp/server/auth/ServerAuthHandlerOverlay.cpp:129]: :0 good catch!

Comment (skyeturriff) [reply src/cpp/server/auth/ServerAuthHandlerOverlay.cpp:129]: 😅 😅 😅 

I was testing the companion changes in Pro, and realized I totally missed the assignment here 🤦🏻‍♀️

**Pull request merged**